packages/typespec-client-generator-core/src/package.ts

```diff
@@ -0,0 +1,734 @@
+import {
+  getClientNamespaceString,
+  getDefaultApiVersion,
+  getLibraryName,
+  getPropertyNames,
+} from "./public-utils.js";
+import {
+  addEncodeInfo,
+  addFormatInfo,
+  getAllModels,
+  getClientType,
+  getSdkCredentialParameter,
+  getSdkModelPropertyType,
+} from "./types.js";
+
+function getSdkHttpBodyParameters(
+  context: SdkContext<SdkHttpOperation>,
+  httpOperation: HttpOperation,
+  methodParameters: SdkParameter[]
+): SdkBodyParameter[] | undefined {
+  const tspBody = httpOperation.parameters.body;
+  if (tspBody === undefined) return undefined;
+  let contentTypes = tspBody.contentTypes;
+  if (contentTypes.length === 0) {
+    contentTypes = ["application/json"];
+  }
+  const defaultContentType = contentTypes.includes("application/json")
+    ? "application/json"
+    : contentTypes[0];
+  if (!tspBody.parameter) {
+    const bodyType = getClientType(context, tspBody.type, httpOperation.operation);
+    return [
+      {
+        kind: "body",
+        nameInClient: "body",
+        description: getDocHelper(context, tspBody.type).description,
+        details: getDocHelper(context, tspBody.type).details,
+        onClient: false,
+        contentTypes: contentTypes,
+        defaultContentType: defaultContentType,
+        isApiVersionParam: false,
+        apiVersions: getAvailableApiVersions<SdkHttpOperation>(context, tspBody.type),
+        type: bodyType,
+        optional: false,
+      },
+    ];
+  }
+  const body = getSdkModelPropertyType(context, tspBody.parameter!, {
+    operation: httpOperation.operation,
+    defaultContentType: defaultContentType,
+  });
+  const methodBodyParameter = methodParameters.find(
+    (x) => x.nameInClient === getPropertyNames(context, tspBody.parameter!)[0]
+  );
+  if (body.kind !== "body") throw new Error("blah");
+  if (methodBodyParameter) {
+    return [
+      {
+        ...body,
+        contentTypes,
+        defaultContentType,
+      },
+    ];
+  } else {
+    // this means that the body parameter is a property on one of the method parameters
+    for (const methodParameter of methodParameters) {
+      if (methodParameter.type.kind === "model") {
+        const bodyProperty = methodParameter.type.properties.find((x) => x.kind === "body");
+        if (bodyProperty) {
+          return [
+            {
+              ...body,
+              contentTypes,
+              defaultContentType,
+            },
+          ];
+        }
+      }
+    }
+  }
+  throw new Error("blah");
+}
+
+function createContentTypeOrAcceptHeader(
+  context: SdkContext<SdkHttpOperation>,
+  bodyObject: SdkBodyParameter | SdkHttpResponse
+): Omit<SdkMethodParameter, "kind"> {
+  const nameInClient = bodyObject.kind === "body" ? "contentType" : "accept";
+  let type: SdkType = {
+    kind: "string",
+    encode: "string",
+    nullable: false,
+  };
+  if (
+    bodyObject.contentTypes &&
+    bodyObject.contentTypes.length === 1 &&
+    /json/.test(bodyObject.contentTypes[0])
```

Comment:
If `length === 1` shouldn't this always be constant? Unclear why it's limited to `json`.

Reply:
For content types, we currently only want to have a constant when it's `json`, i believe this was the behavior as well with `m4`. The reason is bc we don't want to break users. Let's say a service starts off just accepting an `image/png` body, and we generate the `contentType` with constant value `image/png`. If they later add `image/jpeg`, we are now going to require users to input `contentType` automatically during generation, because it's no longer a constant, and between these two, the only clear indicator of what should be defaulted is what was added first, but I feel that's a brittle heuristic since we don't want to get into client default values

Reply:
That makes sense, thanks for the explanation.